sdk/core/core-util/src/delay.ts

```diff
@@ -19,33 +19,71 @@ export interface DelayOptions {
   abortErrorMsg?: string;
 }
 
+/**
+ *
+ * @param inputs - The inputs for creating an abortable promise are the
+ *                 buildPromise function and the cleanupBeforeAbort function.
+ *                 buildPromise takes both the resolve and reject functions as
+ *                 parameters. cleanupBeforeAbort is called right before the
+ *                 promise is rejected.
+ * @returns a function that takes an optional DelayOptions parameter and returns
+ *          a promise that can be aborted.
+ * @internal
+ */
+export function createAbortablePromise<T>(inputs: {
+  buildPromise: (inputs: {
+    resolve: (value: T | PromiseLike<T>) => void;
+    reject: (reason?: any) => void;
+  }) => void;
+  cleanupBeforeAbort?: () => void;
+}): (options?: DelayOptions) => Promise<T> {
+  const { buildPromise, cleanupBeforeAbort } = inputs;
+  return ({ abortSignal, abortErrorMsg } = {}) =>
+    new Promise((resolve, reject) => {
+      function rejectOnAbort(): void {
+        reject(new AbortError(abortErrorMsg ?? "The operation was aborted."));
+      }
+      function removeListeners(): void {
+        abortSignal?.removeEventListener("abort", onAbort);
+      }
+      function onAbort(): void {
+        cleanupBeforeAbort?.();
+        removeListeners();
+        rejectOnAbort();
+      }
+      if (abortSignal?.aborted) {
+        return rejectOnAbort();
+      }
+      buildPromise({
+        resolve: (x) => {
+          resolve(x);
+          removeListeners();
```

Comment:
does the order matter? the old `delay` code removes listeners before resolving.

Reply:
I reverted the ordering but I don't think it matters.